### PR TITLE
Extract board from classic and rename

### PR DESCRIPTION
**Pull request by izakfr**
Closes: #123

### CODE REVIEW

File: server/pkg/models/board/variants/classic.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
+package variants
+
+import (
+	"github.com/variant64/server/pkg/models/board"
+)
+
+// NewClassicBoard creates a new Board with classic rules and returns it.
+func NewClassicBoard() *board.Board {
```
Comment (jbsummoner) [inline server/pkg/models/board/variants/classic.go:8]: 🦸 🧼

File: server/pkg/models/board/board.go [hunk 2]
```diff
@@ -259,3 +259,164 @@ func (b Bounds) IsInboundsPosition(position Position) bool {
...
+			if destination == move.Destination {
+				return nil
+			}
+		}
+	}
+	return errors.New("invalid move: destination position is not a valid destination for the specified move type")
+}
+
+// updateAvailableMoves sets the available moves for each piece in the game.
+func (b *Board) updateAvailableMoves() {
```
Comment (jbsummoner) [inline server/pkg/models/board/board.go:401]: I would decompose this for readability 

```golang
// updateAvailableMoves sets the available moves for each piece in the game.
func (b *Board) updateAvailableMoves() {
	b.forEachPiece(func(rank, file int, piece *Piece) {
		if piece != nil {
			source := Position{Rank: rank, File: file}
			piece.AvailableMoves = b.filterMoveMap(source, piece.GetMoves(source))
		}
	})
}

func (b *Board) forEachPiece(fn func(rank, file int, piece *Piece)) {
	for rank, files := range b.GameboardState {
		for file, piece := range files {
			fn(rank, file, piece)
		}
	}
}

func (b *Board) filterMoveMap(source Position, possibleMoves MoveMap) MoveMap {
	availableMoves := NewMoveMap()
	for moveType, moveListByType := range possibleMoves {
		availableMoves[moveType] = b.filterMoveList(source, moveType, moveListByType)
	}

	return availableMoves
}

func (b *Board) filterMoveList(source Position, moveType MoveType, moveList []Position) []Position {
	availableMoves := []Position{}
	for _, destination := range moveList {
		move := Move{Source: source, Destination: destination, MoveType: moveType}
		if b.IsLegalMove(move, b.GameboardState) {
			availableMoves = append(availableMoves, destination)
		}
	}

	return availableMoves
}
```

File: server/pkg/models/game/actions.go [hunk 3]
```diff
@@ -78,7 +78,7 @@ func (r *RequestNewGame) PerformAction() (*Game, error) {
 func newGameboard(gameboardType board.GameboardType) (gameboard, error) {
```
Comment (jbsummoner) [inline server/pkg/models/game/actions.go:78]: seems more appropriate to be in `/board` as a factory

unless we are taking configs beyond gamboard type

File: server/pkg/models/board/board.go [hunk 4]
```diff
@@ -259,3 +259,164 @@ func (b Bounds) IsInboundsPosition(position Position) bool {
...
+
+type BoardBuilder struct {
+	bounds         Bounds
+	castlingState  *CastlingState
+	moveApplicator *MoveApplicator
+	moveFilter     *MoveFilter
+	gameboardState GameboardState
+}
+
+func NewDefaultBoardBuilder() *BoardBuilder {
```
[the remark this answers is not in the document]
Comment (izakfr) [reply server/pkg/models/board/board.go:273]: done

File: server/pkg/models/board/board.go [hunk 5]
```diff
@@ -259,3 +259,164 @@ func (b Bounds) IsInboundsPosition(position Position) bool {
 		position.Rank < b.RankCount &&
 		position.File < b.FileCount
 }
+
+type builderOption = func(c *BoardBuilder)
+
+type BoardBuilder struct {
```
[the remark this answers is not in the document]
Comment (izakfr) [reply server/pkg/models/board/board.go:265]: done

File: server/pkg/models/board/board.go [hunk 6]
```diff
@@ -259,3 +259,164 @@ func (b Bounds) IsInboundsPosition(position Position) bool {
...
+	}
+}
+
+func WithGameboardState(state GameboardState) builderOption {
+	return func(c *BoardBuilder) {
+		c.gameboardState = NewGameboardState(c.bounds, state)
+	}
+}
+
+func BuildBoard(options ...builderOption) *Board {
```
[the remark this answers is not in the document]
Comment (izakfr) [reply server/pkg/models/board/board.go:330]: done

File: server/pkg/models/board/board.go [hunk 7]
```diff
@@ -259,3 +259,164 @@ func (b Bounds) IsInboundsPosition(position Position) bool {
...
+
+type Board struct {
+	Bounds
+	*MoveApplicator
+	*MoveFilter
+	*CastlingState
+	GameboardState
+}
+
+// GetState returns a GameboardState for the ClassicBoard.
```
[the remark this answers is not in the document]
Comment (izakfr) [reply server/pkg/models/board/board.go:354]: done

File: server/pkg/models/board/board.go (hunk 2 above)
Comment (izakfr) [reply server/pkg/models/board/board.go:401]: this is so clean, thanks!

File: server/pkg/models/game/actions.go (hunk 3 above)
Comment (izakfr) [reply server/pkg/models/game/actions.go:78]: this removes a circular dependency between `variants/` and `board/`

**Review by benshumaker — APPROVED** [review]
I read these 1225 lines of code

**Pull request merged**